pyamg/amg_core/air.h

```diff
@@ -0,0 +1,584 @@
+ *      Cx : const array<float>
+ *          Data array for SOC matrix, C
+ *      Cpts : array<int>
+ *          List of global C-point indices
+ *      splitting : const array<int>
+ *          Boolean array with 1 denoting C-points and 0 F-points
+ *      blocksize : int
+ *          Blocksize of matrix (assume square blocks)
+ *      distance : int, default 2
+ *          Distance of F-point neighborhood to consider, options are 1 and 2.
+ *      use_gmres : bool, default 0
+ *          Use GMRES for local dense solve
+ *      maxiter : int, default 10
+ *          Maximum GMRES iterations
+ *      precondition : bool, default True
+ *          Diagonally precondition GMRES
+ *
+ * Returns
+ * -------
+ * Nothing, Rj[] and Rx[] modified in place.
+ *
+ * Notes
+ * -----
+ * Rx[] must be passed in initialized to zero.
+ */
+template<class I, class T>
+void block_approx_ideal_restriction_pass2(const I Rp[], const int Rp_size,
+                                                I Rj[], const int Rj_size,
+                                                T Rx[], const int Rx_size,
+                                          const I Ap[], const int Ap_size,
+                                          const I Aj[], const int Aj_size,
+                                          const T Ax[], const int Ax_size,
+                                          const I Cp[], const int Cp_size,
+                                          const I Cj[], const int Cj_size,
+                                          const T Cx[], const int Cx_size,
+                                          const I Cpts[], const int Cpts_size,
+                                          const I splitting[], const int splitting_size,
+                                          const I blocksize,
+                                          const I distance = 2,
+                                          const I use_gmres = 0,
+                                          const I maxiter = 10,
+                                          const I precondition = 1 )
+{
+    I is_col_major = true;
+
+    // Build column indices and data for each row of R.
+    for (I row=0; row<Cpts_size; row++) {
+
+        I cpoint = Cpts[row];
+        I ind = Rp[row];
+
+        // Set column indices for R as strongly connected F-points.
+        for (I i=Cp[cpoint]; i<Cp[cpoint+1]; i++) {
+            I this_point = Cj[i];
+            if (splitting[this_point] == F_NODE) {
+                Rj[ind] = Cj[i];
+                ind += 1 ;
+
+                // Strong distance-two F-to-F connections
+                if (distance == 2) {
+                    for (I kk = Cp[this_point]; kk < Cp[this_point+1]; kk++){
+                        if ((splitting[Cj[kk]] == F_NODE) && (this_point != cpoint)) {
+                            Rj[ind] = Cj[kk];
+                            ind += 1 ;
+                        }
+                    } 
+                }
+            }
+        }
+
```

Comment:
What do you think...should this print to cerr or cout, and should it mention the routine triggering the warning?